### DISCUSSION TITLE
SmartSpin2K and qdomyos-zwift

### DISCUSSION THREAD

**Discussion by marklogan17** [doudar/SmartSpin2k#206] [General]
So I have 'almost' built my smartspin2k.  I'm still waiting for the motor from aliexpress, so I caved and ordered one from amazon which will be here Monday.  Everything else is done I think.  So that leads me to the important question...  I helped get qdomyos-zwift (https://github.com/cagnulein/qdomyos-zwift) working with the Echelon bikes.  We have built a ton of functionality in to track data and communicate with all of the other systems out there.  The developer even figured out how to control the resistance in most of the Echelon bikes.  Just my poor Echelon Sport and a couple of others don't have motorized resistance.  So that leads me to my question:  I would love to use QZ in conjunction with the Smartspin, so what would be the best way to do that?  I know you've implemented controllable to set resistance, so is it best just to connect to that?    Ideal I think would be bike->QZ->zwift (or peleton or whatever)->smartspin->bike resistance.  What feedback does smartspin need so that it knows when its at the right resistance or does it just rotate the resistance knob a certain amount?  

There are a LOT of Echelon Sport users in our Facebook group that will jump all over building/buying one of these devices if we solve this in a reasonably simple fashion.  They're all jealous of the ability of the more expensive bikes to have controlled resistance for Peleton in particular :) 

Note - I saw you used ECBHT (can't remember exact name) to decode the echelon protocol.  Its OK for resistance and cadence but its output calculations are not particularly accurate.  You may want to look at the output table and mechanism in QZ to be much more accurate with the Echelon bikes in general.

**Comment by tandaylor** [comment]
Following this thread, Mark.

I'm using the QDDOMYOS app with Zwift and my Echelon Sport. It works amazingly well. Would be great if it could integrate with this device.

I've printed all components and started building the non-pcb model. (I've also ordered 10 boards from pcbway).

Will share my build separately.

**Reply by marklogan17** [reply]
Roberto is already looking at adding this as an FTMS accessory so I should be able to test that (in a motor-less fashion) when he's done.  I think the combination of SmartSpin and QZ will be huge.  We're going on the assumption that changing a level of resistance is essentially a number of motor steps for now.

**Reply by doudar** [reply]
> So I have 'almost' built my smartspin2k. I'm still waiting for the motor from aliexpress, so I caved and ordered one from amazon which will be here Monday. Everything else is done I think. So that leads me to the important question... I helped get qdomyos-zwift (https://github.com/cagnulein/qdomyos-zwift) working with the Echelon bikes. We have built a ton of functionality in to track data and communicate with all of the other systems out there. The developer even figured out how to control the resistance in most of the Echelon bikes. Just my poor Echelon Sport and a couple of others don't have motorized resistance. So that leads me to my question: I would love to use QZ in conjunction with the Smartspin, so what would be the best way to do that? I know you've implemented controllable to set resistance, so is it best just to connect to that? Ideal I think would be bike->QZ->zwift (or peleton or whatever)->smartspin->bike resistance. What feedback does smartspin need so that it knows when its at the right resistance or does it just rotate the resistance knob a certain amount?
> 
> There are a LOT of Echelon Sport users in our Facebook group that will jump all over building/buying one of these devices if we solve this in a reasonably simple fashion. They're all jealous of the ability of the more expensive bikes to have controlled resistance for Peleton in particular :)
> 
> Note - I saw you used ECBHT (can't remember exact name) to decode the echelon protocol. Its OK for resistance and cadence but its output calculations are not particularly accurate. You may want to look at the output table and mechanism in QZ to be much more accurate with the Echelon bikes in general.

Sorry I'm a little slow to respond. Work has been crazy suddenly and I'm also working on a bunch of stuff behind the scenes at the moment for improvements.

As far as feedback from SmartSpin2k, if you don't need ERG mode, sim mode works just fine without anything paired to it. It just needs to be connected to an app as a controllable trainer and it will increase the resistance a set amount (determined by the incline multiplier). 

You're correct, resistance/incline is translated to motor steps here. It really makes sense in simplifying the problem.

If you pair a power source with the SS2K, it will allow ERG mode and adjust the resistance knob to meet whatever watt target the app supplies. 

I'll take a look at QZ. Feel free to do any PR's here :) - Thanks!

**Reply by marklogan17** [reply]
I was talking to Roberto (developer) earlier this week and he took a look at your code and came to the same conclusion.  He originally was thinking of making a change to your code and submitting a PR, but I think he's leaning towards translating within the QZ app and sending the appropriate request.  So for Zwift and such it may be a bit of a pass through depending on mode, but for Peleton he would take the required resistance, know what the bike is currently at and then convert to the correct increase/decrease amount to send here.

My motor comes Monday and knowing Roberto he'll have me testing this pretty quickly.  Will update here with progress.

**Reply by doudar** [reply]
I don't know if it matters much on your end but we're also thinking about implementing a "ftms passthrough" mode where SS2K would talk directly to the app via ftms but then could control another ftms trainer attached to the SS2K receiving the incline or erg target. It would allow things like the "trainer difficulty" setting in Zwift for any app on a trainer other than SS2K, adding more usability to SS2K without building the hardware.

**Reply by marklogan17** [reply]
It would help with Zwift and others that support FTMS.  I don't think it would help with Peleton as communication is via webservice to get required resistance/cadence info.  So for consistency its probably best to try and get it to work with the sim mode.  Seems that my motor may actually arrive today so there's a chance I can actually test my build properly :)

**Reply by marklogan17** [reply]
Just a quick update:  With my build actually 'working' I got to somewhat test the changes the QZ developer has made.  It looks to me like we have the beginnings of a working integration.  He is connecting to SmartSpin as an FTMS accessory and setting the incline value.  Next step will be to get my build all packaged up and installed on the bike so that I can tweak parameters to get it changing resistance appropriately.  We've had some conversations about syncing shifter button presses to QZ as well.  I really think this integration has great implications for both projects!!

**Reply by doudar** [reply]
Make sure to checkout this branch so we’re not doubling efforts too bad. I worked on it a few hours last night. Beginnings of resistance control mode. 

I’m also not opposed to a custom characteristic that provides low level integration either such as direct motor step reading and shifter presses. 
https://github.com/doudar/SmartSpin2k/tree/FTMS_Server-fillout

**Reply by marklogan17** [reply]
Just wanted to come back here and let everyone know that this is now working!!  Thanks @doudar for your work to provide the mechanism to communicate between SS2K and the QZ app.  I did a 10 minute Peloton ride as a test the other night and it followed along the resistance perfectly.  I'm sure we'll hit some edge cases here and there as we use it but it really adds a whole new level to the experience!

**Reply by doudar** [reply]
> Just wanted to come back here and let everyone know that this is now working!! Thanks @doudar for your work to provide the mechanism to communicate between SS2K and the QZ app. I did a 10 minute Peloton ride as a test the other night and it followed along the resistance perfectly. I'm sure we'll hit some edge cases here and there as we use it but it really adds a whole new level to the experience!

Sorry it took a little longer than expected Mark. I needed to take a break (well slowdown at least :) ) from the project for a while. Hitting some issues hard and fast now though.

**Reply by marklogan17** [reply]
Oh no problem at all.  Life is busy for everyone so I don't complain when people are doing extra stuff from the goodness of their heart!!

**Comment by brandankraft** [comment]
I just got my SmartSpin2k running, and I'd like to set it up to work with qz fitness (qdomyoszwift).  I thought it would be simple, but it turns out it isn't... at least not for me.

I've tried doing a couple of things.   

1)   Connecting my Schwinn IC4 spin bike to SmartSpin2K.   Then connecting SmartSpin2K to QZ Fitness.   That didn't get me anywhere.  Cadence and power was not being displayed on QZ Fitness.

Then I tried 

2). Connecting the IC4 to QZ Fitness.   Then adding SmartSpin2k as an accessory within QZ, restarting the app, and starting a peloton ride..  I didn't get any response from smartspin2k on the auto resistance...

So what do I need to do to make this work?  Any ideas?   Any help would be appreciated.

**Reply by eMadman** [reply]
@marklogan17 might be the closest thing to a QZ expert to comment on the SS2K discussion group.  I'll defer to his suggestion.  I know it's a little temperamental to set up with QZ but it should work great once everything is communicating properly.

I can play with it on my bike later this week if you still need a hand.

Worth searching the QZ facebook group as well since I think folks have posted about it.

**Reply by brandankraft** [reply]
Thank you eMadman.... I got it working tonight... I had to delete the QZ app from my device then set up everything again.  I cleared all bluetooth settings from SmartSpin2k.  I connected the IC4 directly to the QZ fitness app, and then made sure to connect SmartSpin2K to QZ as an accessory.  I probably toggled something in the QZ app that messed everything up before.   But everything works great now.   Just finished a peloton ride and it was good!  Now I just need to figure out how to calibrate QZ better...

**Reply by wassa22** [reply]
In qz do you just pick smartspin2k from the devices list. 😁 I thought it showed ftms and wifi when playing around before getting the unit but once I clicked on wifi the other option disappeared.

Cheers

Warwick

**Reply by marklogan17** [reply]
In QZ yes just pick SmartSpin2K, hit OK so it saves and stay with the default shift steps and such for now.  From what you said in Facebook it sounds like you need to get the SS2K config sorted first - get the wifi set up, get the bluetooth stuff disabled, etc.

**Comment by clemco** [comment]
Hi, I have the same setup with IC4, QZ on iphone and Peloton app on ipad. I felt like it was working with a Peloton cycling workout but unfortunately no. It started out fine around 35 but as soon as the instructor called out 40 the SS2K spun up to resistance over 600 and I  pulled it off so it would not destroy the unit. Frankly, I don't really care about the Peloton workouts, and definitely don't want to destroy my SS2K with it, as I find the SS2K with MyWhoosh or Zwift works perfectly. Just pointing it out if there are still bugs out there that need to be fixed.
